py-polars/tests/test_cfg.py

```diff
@@ -66,23 +66,92 @@ def test_tbl_width_chars(environ: None) -> None:
     assert max(len(line) for line in str(df).split("\n")) == 13
 
 
-def test_tbl_cols(environ: None) -> None:
+def test_set_tbl_cols(environ: None) -> None:
     df = pl.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9]})
 
     pl.Config.set_tbl_cols(1)
     assert str(df).split("\n")[2] == "│ a   ┆ ... │"
     pl.Config.set_tbl_cols(2)
-    assert str(df).split("\n")[2] == "│ a   ┆ ... ┆ c   │"
+    assert str(df).split("\n")[2] == "│ a   ┆ b   ┆ c   │"
     pl.Config.set_tbl_cols(3)
     assert str(df).split("\n")[2] == "│ a   ┆ b   ┆ c   │"
 
+    df = pl.DataFrame(
+        {"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9], "d": [10, 11, 12]}
+    )
+    pl.Config.set_tbl_cols(2)
+    assert str(df).split("\n")[2] == "│ a   ┆ ... ┆ d   │"
+    pl.Config.set_tbl_cols(3)
+    assert str(df).split("\n")[2] == "│ a   ┆ b   ┆ c   ┆ d   │"
+
 
-@pytest.mark.skip("not correctly implemented at the moment")
 def test_set_tbl_rows(environ: None) -> None:
-    # df = pl.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9]})
-    pass
-    #
-    # pl.Config.set_tbl_rows(3)
+    df = pl.DataFrame({"a": [1, 2, 3, 4], "b": [5, 6, 7, 8], "c": [9, 10, 11, 12]})
+
+    pl.Config.set_tbl_rows(1)
+    assert (
+        str(df) == "shape: (4, 3)\n"
+        "┌─────┬─────┬─────┐\n"
+        "│ a   ┆ b   ┆ c   │\n"
+        "│ --- ┆ --- ┆ --- │\n"
+        "│ i64 ┆ i64 ┆ i64 │\n"
+        "╞═════╪═════╪═════╡\n"
+        "│ 1   ┆ 5   ┆ 9   │\n"
+        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
+        "│ ... ┆ ... ┆ ... │\n"
+        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
+        "│ 4   ┆ 8   ┆ 12  │\n"
+        "└─────┴─────┴─────┘"
+    )
+    pl.Config.set_tbl_rows(2)
+    assert (
+        str(df) == "shape: (4, 3)\n"
+        "┌─────┬─────┬─────┐\n"
+        "│ a   ┆ b   ┆ c   │\n"
+        "│ --- ┆ --- ┆ --- │\n"
+        "│ i64 ┆ i64 ┆ i64 │\n"
+        "╞═════╪═════╪═════╡\n"
+        "│ 1   ┆ 5   ┆ 9   │\n"
+        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
+        "│ ... ┆ ... ┆ ... │\n"
+        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
+        "│ 4   ┆ 8   ┆ 12  │\n"
+        "└─────┴─────┴─────┘"
+    )
+    pl.Config.set_tbl_rows(3)
```

Comment:
It seems to be incorrect? We set 3 rows and we get 4?

Reply:
I have modified the behaviour to display the full DataFrame when we set `set_tbl_rows` or `set_tbl_cols` to be `height-1` or `width-1`

Old:
```python
        "┌─────┬─────┬─────┐\n"
        "│ a   ┆ b   ┆ c   │\n"
        "│ --- ┆ --- ┆ --- │\n"
        "│ i64 ┆ i64 ┆ i64 │\n"
        "╞═════╪═════╪═════╡\n"
        "│ 1   ┆ 5   ┆ 9   │\n"
        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
        "│ 2   ┆ 6   ┆ 10  │\n"
        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
        "│ ... ┆ ... ┆ ... │\n"
        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
        "│ 4   ┆ 8   ┆ 12  │\n"
        "└─────┴─────┴─────┘"
```

New:

```python
        "┌─────┬─────┬─────┐\n"
        "│ a   ┆ b   ┆ c   │\n"
        "│ --- ┆ --- ┆ --- │\n"
        "│ i64 ┆ i64 ┆ i64 │\n"
        "╞═════╪═════╪═════╡\n"
        "│ 1   ┆ 5   ┆ 9   │\n"
        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
        "│ 2   ┆ 6   ┆ 10  │\n"
        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
        "│ 3   ┆ 7   ┆ 11  │\n"
        "├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤\n"
        "│ 4   ┆ 8   ┆ 12  │\n"
        "└─────┴─────┴─────┘"
```

As you can see, in both cases 4 (height) rows will be printed. In the old behaviour, we would have `...` rows which are not useful, might as well show the full Dataframe since the string table will ultimately have 4  rows.  

This only happens when `set_tbl_rows(height-1)` or `set_tbl_cols(width-1)`. If we were to `set_tbl_rows(3)` on a DataFrame with `height >= 5` we will get a string with `4 rows (3 data row + 1 ...)`

I can revert back to the old behaviour if that's better.

Reply:
Note:

FAIL: over (polars.internals.expr.Expr)
Doctest: polars.internals.expr.Expr.over

FAIL: to_dummies (polars.internals.frame.DataFrame)
Doctest: polars.internals.frame.DataFrame.to_dummies

Failing due to the above new behaviour

Reply:
I agree with @ritchie46  that this is not intuitive, I actually made the same remark for the cols.

Reply:
Ah, I see. 

Due to the `...` row a single row, doesn't make sense, so will always be 3, so that we can insert the `...` row.

I now understand the 1 off error.  Maybe we can adapt the test so that we test on 5 or 6 rows as well, then its less counter intuitive. And then let's add a comment to the test where we show 4 rows instead of 3 (so that we don't forget in the future).

Reply:
I think it's best to keep it the way it is, avoiding confusion.
I have added a test case with 5 rows.